x-template.js

```diff
@@ -269,6 +270,1104 @@ class Forgiving {
+  ]);
+
+  static #allowedMathElements = new Set([
+    // Top-level elements
+    // https://developer.mozilla.org/en-US/docs/Web/MathML/Element#top-level_elements
+    'math',
+
+    // Token elements
+    // https://developer.mozilla.org/en-US/docs/Web/MathML/Element#token_elements
+    'mi', 'mn', 'mo', 'ms', 'mspace', 'mtext',
+
+    // General layout
+    // https://developer.mozilla.org/en-US/docs/Web/MathML/Element#general_layout
+    'merror', 'mfrac', 'mpadded', 'mphantom', 'mroot', 'mrow',
+    'msqrt', 'mstyle',
+    // 'menuclose', 'mfenced',
+
+    // Script and limit elements
+    // https://developer.mozilla.org/en-US/docs/Web/MathML/Element#script_and_limit_elements
+    'mmultiscripts', 'mover', 'mprescripts', 'msub', 'msubsup', 'msup',
+    'munder', 'munderover',
+
+    // Tabular math
+    // https://developer.mozilla.org/en-US/docs/Web/MathML/Element#tabular_math
+    'mtable', 'mtd', 'mtr',
+
+    // Uncategorized elements
+    // https://developer.mozilla.org/en-US/docs/Web/MathML/Element#uncategorized_elements
+    'annotation', 'semantics',
+    // 'annotation-xml',
+
+  ]);
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Parsing State Values //////////////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  // These are all the states we can be in while we parse a string.
+  // Reference: https://w3c.github.io/html-reference/syntax.html
+
+  // The “initial” and “boundContent” states are special in that there is no
+  // related pattern to match. Initial is just the state we start in and we only
+  // find bound content at string terminals (i.e., interpolations).
+  static #initial =      / /y;
+  static #boundContent = / /y;
+
+  // We restrict text to any string of characters that doesn’t cross through a
+  // “<” character.
+  static #unboundContent = /[^<]+/y;
+
+  // Comments start with “<!--” and go until the next occurrence of “-->”. Note
+  // that we use non-greedy matching and we allow the “.” to match newlines.
+  static #unboundComment = /<!--.*?-->/ys;
+
+  // Tag names can contain the lowercase characters [a-z], [0-9], and a hyphen.
+  // But, the first character must be [a-z] and the last character cannot be a
+  // hyphen. The first character of the tag must directly follow the opening
+  // angle bracket. The closing angle bracket of the open tag cannot be preceded
+  // by a space or newline. The closing tag must not contain spaces or newlines.
+  // - ok: <h6>, <my-element-1>
+  // - not ok: <-div>, <1-my-element>
+  static #openTagStart = /<(?![0-9-])[a-z0-9-]+(?<!-)(?=[\s\n>])/y;
+  static #closeTag =   /<\/(?![0-9-])[a-z0-9-]+(?<!-)>/y;
+  static #openTagSpace =   /[\s\n]+/y;
+  static #openTagEnd = /(?<![\s\n])>/y;
+
+  // Attribute names can contain the characters [a-z], [A-Z], [0-9], and a
+  // hyphen. But, they cannot begin with numbers, capital letters, or hyphens —
```

Comment:
question:
>they cannot begin with numbers, capital letters, or hyphens
etc — according to us or w3c?

Reply:
nit: please note the disparity between the w3c spec standard and our own choices

Reply:
I did my best to capture this detail — I’m sure I’ve not done a perfect job yet, but we may need to simply iterate here…